File: lib/screens/login/LoginScreen.java
import 'package:flutter/material.dart';
import '../home/home_screen.dart';
import 'register_screen.dart';

class LoginScreen extends StatelessWidget {
  const LoginScreen({super.key});

  @override
  Widget build(BuildContext context) {
    final bool isSmallScreen = MediaQuery.of(context).size.width < 600;

    return Scaffold(
      body: Center(
        child: isSmallScreen
            ? const Column(
          mainAxisSize: MainAxisSize.min,
          children: [_Logo(), _FormContent()],
        )
            : Container(
          padding: const EdgeInsets.all(32.0),
          constraints: const BoxConstraints(maxWidth: 800),
          child: const Row(
            children: [
              Expanded(child: _Logo()),
              Expanded(child: Center(child: _FormContent())),
            ],
          ),
        ),
      ),
    );
  }
}

class _Logo extends StatelessWidget {
  const _Logo();

  @override
  Widget build(BuildContext context) {
    final bool isSmallScreen = MediaQuery.of(context).size.width < 600;

    return Column(
      mainAxisSize: MainAxisSize.min,
      children: [
        Icon(Icons.local_hospital, size: isSmallScreen ? 100 : 200, color: Colors.blue),
        Padding(
          padding: const EdgeInsets.all(16.0),
          child: Text(
            "Bienvenido a Clínica Innova",
            textAlign: TextAlign.center,
            style: isSmallScreen
                ? Theme.of(context).textTheme.titleMedium
                : Theme.of(context).textTheme.titleSmall,
          ),
        ),
      ],
    );
  }
}

class _FormContent extends StatefulWidget {
  const _FormContent();

  @override
  State<_FormContent> createState() => __FormContentState();
}

class __FormContentState extends State<_FormContent> {
  bool _isPasswordVisible = false;
  bool _rememberMe = false;

  final GlobalKey<FormState> _formKey = GlobalKey<FormState>();

  @override
  Widget build(BuildContext context) {
    return Container(
      constraints: const BoxConstraints(maxWidth: 300),
      child: Form(
        key: _formKey,
        child: Column(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            // Email
            TextFormField(
              validator: (value) {
                if (value == null || value.isEmpty) {
                  return 'Por favor, introduce tu correo';
                }
                bool emailValid = RegExp(
                  r"^[a-zA-Z0-9.a-zA-Z0-9.!#$%&'*+-/=?^_`{|}~]+@[a-zA-Z0-9]+\.[a-zA-Z]+",
                ).hasMatch(value);
                if (!emailValid) {
                  return 'Introduce un correo válido';
                }
                return null;
              },
              decoration: const InputDecoration(
                labelText: 'Correo electrónico',
                hintText: 'Introduce tu correo',
                prefixIcon: Icon(Icons.email_outlined),
                border: OutlineInputBorder(),
              ),
            ),
            const SizedBox(height: 16),
            // Password
            TextFormField(
              validator: (value) {
                if (value == null || value.isEmpty) {
                  return 'Por favor, introduce tu contraseña';
                }
                if (value.length < 6) {
                  return 'La contraseña debe tener al menos 6 caracteres';
                }
                return null;
              },
              obscureText: !_isPasswordVisible,
              decoration: InputDecoration(
                labelText: 'Contraseña',
                hintText: 'Introduce tu contraseña',
                prefixIcon: const Icon(Icons.lock_outline_rounded),
                border: const OutlineInputBorder(),
                suffixIcon: IconButton(
                  icon: Icon(
                    _isPasswordVisible ? Icons.visibility_off : Icons.visibility,
                  ),
                  onPressed: () {
                    setState(() {
                      _isPasswordVisible = !_isPasswordVisible;
                    });
                  },
                ),
              ),
            ),
            CheckboxListTile(
              value: _rememberMe,
              onChanged: (value) {
                if (value == null) return;
                setState(() {
                  _rememberMe = value;
                });
              },
              title: const Text('Recordarme'),
              controlAffinity: ListTileControlAffinity.leading,
              dense: true,
              contentPadding: const EdgeInsets.all(0),
            ),
            const SizedBox(height: 16),
            // Login button
            SizedBox(
              width: double.infinity,
              child: ElevatedButton(
                style: ElevatedButton.styleFrom(
                  shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(4)),
                ),
                child: const Padding(
                  padding: EdgeInsets.all(10.0),
                  child: Text(
                    'Iniciar sesión',
                    style: TextStyle(fontSize: 16, fontWeight: FontWeight.bold),
                  ),
                ),
                onPressed: () {
                  if (_formKey.currentState?.validate() ?? false) {
                    Navigator.of(context).pushReplacement(
                      MaterialPageRoute(builder: (context) => const HomeScreen()),
                    );
                  }
                },
              ),
            ),
            const SizedBox(height: 12),
            // Register button
            TextButton(
              onPressed: () {
                Navigator.of(context).push(
                  MaterialPageRoute(builder: (context) => const RegisterScreen()),
                );
              },
              child: const Text('¿No tienes cuenta? Regístrate'),
            ),
            const SizedBox(height: 12),
            const Text('O inicia sesión con'),
            const SizedBox(height: 12),
            // Social login buttons with Wrap to avoid overflow
            Wrap(
              alignment: WrapAlignment.center,
              spacing: 12,
              runSpacing: 12,
              children: [
                _SocialLoginButton(
                  imageUrl:
                  'https://upload.wikimedia.org/wikipedia/commons/thumb/5/53/Google_%22G%22_Logo.svg/512px-Google_%22G%22_Logo.svg.png',
                  label: 'Google',
                  onPressed: () {
                    // Lógica para iniciar sesión con Google
                  },
                ),
                _SocialLoginButton(
                  imageUrl:
                  'https://upload.wikimedia.org/wikipedia/commons/thumb/f/fa/Apple_logo_black.svg/512px-Apple_logo_black.svg.png',
                  label: 'Apple',
                  onPressed: () {
                    // Lógica para iniciar sesión con Apple
                  },
                ),
                _SocialLoginButton(
                  imageUrl:
                  'https://upload.wikimedia.org/wikipedia/commons/0/05/Facebook_Logo_(2019).png',
                  label: 'Facebook',
                  onPressed: () {
                    // Lógica para iniciar sesión con Facebook
                  },
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

class _SocialLoginButton extends StatelessWidget {
  final String imageUrl;
  final String label;
  final VoidCallback onPressed;

  const _SocialLoginButton({
    required this.imageUrl,
    required this.label,
    required this.onPressed,
  });

  @override
  Widget build(BuildContext context) {
    return OutlinedButton(
      onPressed: onPressed,
      style: OutlinedButton.styleFrom(
        padding: const EdgeInsets.symmetric(vertical: 8, horizontal: 12),
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(6)),
      ),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Image.network(imageUrl, height: 32, width: 32),
          const SizedBox(height: 6),
          Text(label),
        ],
      ),
    );
  }
}

File: lib/screens/login/login_screen.dart
import 'package:flutter/material.dart';
import '../home/home_screen.dart';
import '../onboarding/onboarding_screen.dart' show AnimatedBackground;
import 'register_screen.dart';
import 'forgot_password_screen.dart';

class LoginScreen extends StatefulWidget {
  const LoginScreen({Key? key}) : super(key: key);

  @override
  State<LoginScreen> createState() => _LoginScreenState();
}

class _LoginScreenState extends State<LoginScreen> {
  final _formKey = GlobalKey<FormState>();
  final _emailController = TextEditingController();
  final _passwordController = TextEditingController();
  bool _isPasswordVisible = false;
  bool _rememberMe = false;

  // Definir estilos comunes para reutilizar
  final _whiteColor = Colors.white;
  final _whiteColorWithOpacity = Colors.white.withOpacity(0.7);

  // Optimización de decoraciones reutilizables
  late final InputDecoration _inputDecoration;

  @override
  void initState() {
    super.initState();
    _inputDecoration = InputDecoration(
      border: OutlineInputBorder(
        borderRadius: BorderRadius.circular(12),
        borderSide: BorderSide(color: _whiteColorWithOpacity),
      ),
      enabledBorder: OutlineInputBorder(
        borderRadius: BorderRadius.circular(12),
        borderSide: BorderSide(color: _whiteColorWithOpacity),
      ),
      focusedBorder: OutlineInputBorder(
        borderRadius: BorderRadius.circular(12),
        borderSide: BorderSide(color: _whiteColor),
      ),
      labelStyle: TextStyle(color: _whiteColor),
      hintStyle: TextStyle(color: _whiteColorWithOpacity),
    );
  }

  @override
  void dispose() {
    _emailController.dispose();
    _passwordController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final primaryColor = Theme.of(context).primaryColor;

    return Scaffold(
      extendBodyBehindAppBar: true,
      appBar: AppBar(
        backgroundColor: Colors.transparent,
        elevation: 0,
      ),
      body: Stack(
        children: [
          // Fondo animado
          const AnimatedBackground(),

          // Contenido
          Center(
            child: SingleChildScrollView(
              padding: const EdgeInsets.all(24.0),
              child: Container(
                padding: const EdgeInsets.all(24.0),
                decoration: BoxDecoration(
                  color: Colors.black.withOpacity(0.2),
                  borderRadius: BorderRadius.circular(16),
                ),
                child: Form(
                  key: _formKey,
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    crossAxisAlignment: CrossAxisAlignment.stretch,
                    children: [
                      // Logo o icono
                      Icon(
                        Icons.medical_services,
                        size: 64,
                        color: _whiteColor,
                      ),
                      const SizedBox(height: 24),

                      // Título
                      Text(
                        'Clínica Médica',
                        style: Theme.of(context).textTheme.headlineMedium?.copyWith(
                          fontWeight: FontWeight.bold,
                          color: _whiteColor,
                        ),
                        textAlign: TextAlign.center,
                      ),
                      const SizedBox(height: 8),

                      // Subtítulo
                      Text(
                        'Inicia sesión para continuar',
                        style: Theme.of(context).textTheme.bodyMedium?.copyWith(
                          color: _whiteColor.withOpacity(0.8),
                        ),
                        textAlign: TextAlign.center,
                      ),
                      const SizedBox(height: 32),

                      // Botones de inicio de sesión social
                      Row(
                        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                        children: [
                          _buildSocialLoginButton(
                            icon: 'assets/icons/google.png',
                            fallbackIcon: Icons.g_mobiledata,
                            label: 'Google',
                            onPressed: () => _loginWithSocial(() => const HomeScreen()),
                          ),
                          _buildSocialLoginButton(
                            icon: 'assets/icons/apple.png',
                            fallbackIcon: Icons.apple,
                            label: 'Apple',
                            onPressed: () => _loginWithSocial(() => const HomeScreen()),
                          ),
                          _buildSocialLoginButton(
                            icon: 'assets/icons/facebook.png',
                            fallbackIcon: Icons.facebook,
                            label: 'Facebook',
                            onPressed: () => _loginWithSocial(() => const HomeScreen()),
                          ),
                        ],
                      ),
                      const SizedBox(height: 24),

                      // Separador
                      Row(
                        children: [
                          Expanded(
                            child: Divider(
                              color: _whiteColor.withOpacity(0.5),
                              thickness: 1,
                            ),
                          ),
                          Padding(
                            padding: const EdgeInsets.symmetric(horizontal: 16),
                            child: Text(
                              'O inicia sesión con',
                              style: TextStyle(
                                color: _whiteColor.withOpacity(0.8),
                                fontSize: 12,
                              ),
                            ),
                          ),
                          Expanded(
                            child: Divider(
                              color: _whiteColor.withOpacity(0.5),
                              thickness: 1,
                            ),
                          ),
                        ],
                      ),
                      const SizedBox(height: 24),

                      // Campo de email
                      TextFormField(
                        controller: _emailController,
                        decoration: _inputDecoration.copyWith(
                          labelText: 'Correo electrónico',
                          hintText: 'ejemplo@correo.com',
                          prefixIcon: Icon(Icons.email_outlined, color: _whiteColorWithOpacity),
                        ),
                        style: TextStyle(color: _whiteColor),
                        keyboardType: TextInputType.emailAddress,
                        validator: (value) {
                          return null;
                        },
                      ),
                      const SizedBox(height: 16),

                      // Campo de contraseña
                      TextFormField(
                        controller: _passwordController,
                        decoration: _inputDecoration.copyWith(
                          labelText: 'Contraseña',
                          hintText: 'Ingresa tu contraseña',
                          prefixIcon: Icon(Icons.lock_outline, color: _whiteColorWithOpacity),
                          suffixIcon: IconButton(
                            icon: Icon(
                              _isPasswordVisible ? Icons.visibility_off : Icons.visibility,
                              color: _whiteColorWithOpacity,
                            ),
                            onPressed: () {
                              setState(() {
                                _isPasswordVisible = !_isPasswordVisible;
                              });
                            },
                          ),
                        ),
                        style: TextStyle(color: _whiteColor),
                        obscureText: !_isPasswordVisible,
                        validator: (value) {
                          return null;
                        },
                      ),
                      const SizedBox(height: 8),

                      // Opciones adicionales
                      Row(
                        children: [
                          // Checkbox "Recordarme"
                          Theme(
                            data: ThemeData(
                              unselectedWidgetColor: _whiteColorWithOpacity,
                            ),
                            child: Checkbox(
                              value: _rememberMe,
                              checkColor: Colors.black,
                              activeColor: _whiteColor,
                              onChanged: (value) {
                                setState(() {
                                  _rememberMe = value ?? false;
                                });
                              },
                            ),
                          ),
                          Text('Recordarme', style: TextStyle(color: _whiteColor)),
                          const Spacer(),
                          // Olvidé mi contraseña
                          TextButton(
                            onPressed: () {
                              Navigator.of(context).push(
                                MaterialPageRoute(
                                  builder: (context) => const ForgotPasswordScreen(),
                                ),
                              );
                            },
                            child: Text(
                              '¿Olvidaste tu contraseña?',
                              style: TextStyle(color: _whiteColor),
                            ),
                          ),
                        ],
                      ),
                      const SizedBox(height: 24),

                      // Botón de inicio de sesión
                      ElevatedButton(
                        onPressed: () {
                          // Navegamos directamente a Home sin validar
                          Navigator.of(context).pushReplacement(
                            MaterialPageRoute(
                              builder: (context) => const HomeScreen(),
                            ),
                          );
                        },
                        style: ElevatedButton.styleFrom(
                          padding: const EdgeInsets.symmetric(vertical: 16),
                          backgroundColor: _whiteColor,
                          foregroundColor: primaryColor,
                          shape: RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(12),
                          ),
                        ),
                        child: const Text(
                          'Iniciar Sesión',
                          style: TextStyle(
                            fontSize: 16,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                      ),
                      const SizedBox(height: 16),

                      // Opción para registrarse
                      Row(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          Text('¿No tienes una cuenta?', style: TextStyle(color: _whiteColor)),
                          TextButton(
                            onPressed: () {
                              // Navegar a pantalla de registro
                              Navigator.of(context).push(
                                MaterialPageRoute(
                                  builder: (context) => const RegisterScreen(),
                                ),
                              );
                            },
                            child: Text(
                              'Regístrate',
                              style: TextStyle(
                                color: _whiteColor,
                                fontWeight: FontWeight.bold,
                              ),
                            ),
                          ),
                        ],
                      ),
                    ],
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }

  // Método para optimizar la navegación con inicio de sesión social
  void _loginWithSocial(Widget Function() destinationBuilder) {
    Navigator.of(context).pushReplacement(
      MaterialPageRoute(
        builder: (context) => destinationBuilder(),
      ),
    );
  }

  Widget _buildSocialLoginButton({
    required String icon,
    required IconData fallbackIcon,
    required String label,
    required VoidCallback onPressed,
  }) {
    return Column(
      children: [
        InkWell(
          onTap: onPressed,
          borderRadius: BorderRadius.circular(50),
          child: Container(
            width: 55,
            height: 55,
            decoration: BoxDecoration(
              color: _whiteColor,
              borderRadius: BorderRadius.circular(50),
              boxShadow: [
                BoxShadow(
                  color: Colors.black.withOpacity(0.1),
                  blurRadius: 10,
                  spreadRadius: 1,
                ),
              ],
            ),
            child: Center(
              child: _buildSocialIcon(icon, fallbackIcon),
            ),
          ),
        ),
        const SizedBox(height: 8),
        Text(
          label,
          style: TextStyle(
            color: _whiteColor,
            fontSize: 12,
          ),
        ),
      ],
    );
  }

  Widget _buildSocialIcon(String assetPath, IconData fallbackIcon) {
    // Precache de imágenes para mejorar rendimiento
    final primaryColor = Theme.of(context).primaryColor;

    try {
      return Image.asset(
        assetPath,
        width: 30,
        height: 30,
        errorBuilder: (context, error, stackTrace) {
          return Icon(
            fallbackIcon,
            color: primaryColor,
            size: 30,
          );
        },
      );
    } catch (e) {
      return Icon(
        fallbackIcon,
        color: primaryColor,
        size: 30,
      );
    }
  }
}
